### PR TITLE
 use llir/llvm@v0.3.0-pre2 for code generation

### PR DESCRIPTION
**Pull request by mewmew**
As an experiment, use the latest version of llir/llvm for code generation.

`codegen` test case output:

```
u@x1 ~/D/l/e/codegen> go test -v
=== RUN   TestBinaryFunction
```
```llvm
define i32 @main() {
; <label>:0
	%1 = call i32 @add(i32 10, i32 10)
	ret i32 %1
}

define i32 @add(i32 %x, i32 %y) {
; <label>:0
	%1 = add i32 10, 10
	ret i32 %1
}
```
```
--- PASS: TestBinaryFunction (0.00s)
PASS
ok  	github.com/elz-lang/elz/codegen	0.001s
```

Updates #5.

P.S. the test case was also updated to include a `ret` instruction for the `@main` function.

### CODE REVIEW

File: codegen/codegen_test.go [hunk 1]
```diff
@@ -36,5 +36,6 @@ func TestBinaryFunction(t *testing.T) {
 			},
 		},
 	)
+	builder.NewRet(v)
```
Comment (mewmew) [inline codegen/codegen_test.go]: Note, this ret instruction was added in the PR.

File: codegen/codegen_test.go [hunk 2]
```diff
@@ -20,13 +22,11 @@ func TestBinaryFunction(t *testing.T) {
...
-	ft := llvm.FunctionType(llvm.Int32Type(), []llvm.Type{}, false)
-	mainFn := llvm.AddFunction(c.mod, "main", ft)
-	bb := llvm.AddBasicBlock(mainFn, "")
-	builder := llvm.NewBuilder()
-	builder.SetInsertPointAtEnd(bb)
-	c.NewExpr(
-		map[string]llvm.Value{},
+	mainFn := c.mod.NewFunc("main", types.I32)
+	bb := mainFn.NewBlock("")
+	builder := bb
```
Comment (dannypsnl) [inline codegen/codegen_test.go]: I think here should directly write `mainBuilder := mainFn.NewBlock()`

File: codegen/codegen.go [hunk 3]
```diff
@@ -30,31 +34,29 @@ func (c *CodeGenerator) GenBinding(binding *ast.Binding) {
...
 		}
 		retT := c.GetExprType(scope, binding.Expr)
-		ft := llvm.FunctionType(retT, typeList, false)
-		newFn := llvm.AddFunction(c.mod, binding.Name, ft)
-		newFunctionBuilder := llvm.NewBuilder()
-		block := llvm.AddBasicBlock(newFn, "")
-		newFunctionBuilder.SetInsertPointAtEnd(block)
+		newFn := c.mod.NewFunc(binding.Name, retT, params...)
+		block := newFn.NewBlock("")
+		newFunctionBuilder := block
```
Comment (dannypsnl) [inline codegen/codegen.go]: `newFnBuilder := newFn.NewBlock("")`

**Review by dannypsnl — CHANGES_REQUESTED** [review]

File: codegen/codegen.go (hunk 3 above)
Comment (mewmew) [reply codegen/codegen.go]: > `newFnBuilder := newFn.NewBlock("")`

Done.

Comment (mewmew) [reply codegen/codegen.go]: I made the update, but after thinking about it further, there is a benefit with having the basic block created explicitly (as was done in the original code). The reason is that you may want to update how the builder works, right now, it always uses `SetInsertPointAtEnd` by default.

In particular, I would imagine that the [type Builder interface](https://github.com/mewpull/elz/blob/6d3a3e1c1004e2a57ec9052d39cdf7cb08ac1071/codegen/builder.go#L11) in builder.go would be replaced with a concrete builder implementation that allows for more flexible control over insertion points for instructions.

This is just a suggestion however, and you can update the code at that time, should the need arise.

Cheers,
Robin

Comment (dannypsnl) [reply codegen/codegen.go]: Thanks for the explanation, I'm also thinking why not use concrete type XD.

**Review by dannypsnl — APPROVED** [review]

**Pull request merged**